perf: optimized request headers

AlliBalliBaba opened the pull request:
I noticed that an average browser request actually has quite a few request headers. A request to this github page when logged in will have ~15 headers totaling ~2kb in data.

### Header keys

Currently there's an optimisation in place that uses [otter](https://github.com/maypok86/otter) to cache header keys. After looking at flamegraphs, I noticed that in this case caching is actually not much more efficient than just concatenating the string.
This PR gets rid of the otter dependency and instead caches 80 of the most common request headers directly as a `zend_string` (by explicitly writing them into a map).

### Cookie Header

Additionally, this PR removes the cookie sanitizing on the go side. I noticed that other SAPIs [1](https://github.com/php/php-src/blob/7dbfacb27e7807a924792deaaabb8ab8c228484a/sapi/litespeed/lsapi_main.c#L304) [2](https://github.com/php/php-src/blob/7dbfacb27e7807a924792deaaabb8ab8c228484a/sapi/apache2handler/sapi_apache2.c#L244) [3](https://github.com/php/php-src/blob/7dbfacb27e7807a924792deaaabb8ab8c228484a/sapi/cgi/cgi_main.c#L653) will also just forward the cookie header and let PHP sanitize it.
When testing with Nginx/FPM I also noticed that nginx will even forward null bytes in the cookie, leading to a truncated cookie. I'm pretty sure this will behave the same way on Caddy/FPM, though I haven't explicitly tested there.

Sanitizing the cookie on the go side is actually pretty slow and leads to a slightly different behaviour. Slightly malformed cookies like this would be sanitized differently by GO and PHP:

Cookie: `key =value`
GO trims the key on both sides:  `key=value`
PHP replaces the space with `_`: `key_=value`

PHP will request the cookie from the Sapi everytime it discovers the Cookie header in the request, even if the Cookie is not explicitly read by the PHP script.

### Flamegraphs

Header registration with the 15 github headers:
registering headers consumes ~8.5% of the request, sanitizing the cookie consumes ~9.5%
![flame-inefficient-headers](https://github.com/user-attachments/assets/7273b772-c4d2-4c83-8629-74ae1365ea4d)

Header registration this PR:
registering headers consumes ~4.5% of the request, forwarding the cookie header consumes ~0.5%

![flame-efficient-headers](https://github.com/user-attachments/assets/4442d53b-2d98-4a1f-8473-819b932dc8b9)

withinboredom commented on cgi.go:
```diff
@@ -194,27 +194,38 @@ func getKnownVariableKeys(thread *phpThread) map[string]*C.zend_string {
 
+func getCachedHeaderKey(thread *phpThread, key string) *C.zend_string {
+	if thread.knownHeaderKeys == nil {
+		thread.knownHeaderKeys = make(map[string]*C.zend_string)
```
zend_strings have their own memory lifetime and GC policies in PHP, so we probably need to ensure that the engine doesn't consider this a leak when shutting down the thread (ensure php is running in debug mode).

AlliBalliBaba replied:
I was manually releasing these strings before, but then I realized that they get released automatically [on thread shutdown](https://github.com/dunglas/frankenphp/blob/92e6b48156ae029d96d2fac8c462b94138e4dab1/frankenphp.c#L905) together with all other TSRM resources. So we just need to make sure go doesn't keep any pointers to them.

withinboredom commented on cgi.go:
```diff
@@ -194,27 +194,38 @@ func getKnownVariableKeys(thread *phpThread) map[string]*C.zend_string {
+		return nil
+	}
+
+	zendHeader := C.frankenphp_init_persistent_string(toUnsafeChar(commonKey), C.size_t(len(commonKey)))
```
I tested this on a draft RFC implementation that I worked on for a client, and it seems that if it is accepted, using persistent strings like this will cause issues in php 8.5. This isn't an issue right now (obviously), but generally, persistent strings are used by the compiler and ini entries, and not headers (user input). PHP generally considers persistent strings to be "safe".

AlliBalliBaba replied:
The keys are definitely safe in this case since they are only those hard-coded in `headers.go`. I was wondering before if there are better ways to create strings that are ignored by the GC. Ideally they would just be registered once on the main thread, but it looks like that causes segfaults in ZTS.
Let's keep this in mind, it's the same as for the $_SERVER keys and it shouldn't be hard to update the implementation

withinboredom replied:
Hmmm. `GC_NOT_COLLECTABLE` _might_ work, but I don't think it is tested on zend strings. You also might want to set it as `GC_PERMANENT` (ie, `zend_add_interned_string`) to prevent in-place modification optimizations from changing the header.

withinboredom replied:
In any case, it isn't a blocker here.

withinboredom replied:
I also am relatively certain that interned strings have to be configured before/during engine initialization.

AlliBalliBaba replied:
I'll try and see how it behaves with interned strings 👍

withinboredom commented on internal/headers/headers.go:
```diff
@@ -0,0 +1,103 @@
+	"Authorization":             "HTTP_AUTHORIZATION",
+	"Cache-Control":             "HTTP_CACHE_CONTROL",
+	"Connection":                "HTTP_CONNECTION",
+	"Content-Length":            "HTTP_CONTENT_LENGTH",
```
I believe this should be `CONTENT_LENGTH` and not `HTTP_CONTENT_LENGTH`?

https://www.php.net/manual/en/reserved.variables.server.php#:~:text=If%20requests%20to,%24_SERVER%5B%27CONTENT_TYPE%27%5D

AlliBalliBaba replied:
`$_SERVER` will actually have both `CONTENT_LENGTH `and `HTTP_CONTENT_LENGTH`, `CONENT_LENGTH` is additionally registered [here](https://github.com/dunglas/frankenphp/blob/92e6b48156ae029d96d2fac8c462b94138e4dab1/cgi.go#L134)

AlliBalliBaba replied:
I just tested this with nginx/FPM and it will also not remove `HTTP_CONTENT_LENGTH` and `HTTP_CONTENT_TYPE` from `$_SERVER` contrary to what the linked comment says (the orginal rfc link sadly leads to a 404 page)

withinboredom replied:
😂

withinboredom commented on internal/headers/headers.go:
```diff
@@ -0,0 +1,103 @@
+	"Referer":                   "HTTP_REFERER",
+	"Cookie":                    "HTTP_COOKIE",
+	"Host":                      "HTTP_HOST",
+	"Content-Type":              "HTTP_CONTENT_TYPE",
```
Same here?

AlliBalliBaba replied:
Same as above, `HTTP_CONTENT_TYPE` and `CONTENT_TYPE` will both be present in the final `$_SERVER` array. I'll explicitly add this to the server vars test.

withinboredom commented on internal/headers/headers.go:
```diff
@@ -0,0 +1,103 @@
+	"Transfer-Encoding":         "HTTP_TRANSFER_ENCODING",
+	"Upgrade":                   "HTTP_UPGRADE",
+	"DNT":                       "HTTP_DNT",
+	"Authorization-Bearer":      "HTTP_AUTHORIZATION_BEARER",
```
I have never seen this header before, is this correct?

AlliBalliBaba replied:
Tbh, "Authorization-Bearer" might also have been hallucinated by ChatGPT since I asked it for additional headers. I'll remove it

withinboredom replied:
Gotta love AI -- making things harder and easier at the same time!

withinboredom commented on internal/headers/headers.go:
```diff
@@ -0,0 +1,103 @@
+	"Warning":             "HTTP_WARNING",
+	"Priority":            "HTTP_PRIORITY",
+	"TE":                  "HTTP_TE",
+	"Trailer":             "HTTP_TRAILER",
```
Might also be good to include `Etag`, `Range`, and maybe `Clear-Site-Data`.

AlliBalliBaba replied:
Will do 👍

withinboredom reviewed (approved):
Some minor issues with the list but LGTM otherwise.

dunglas commented:
Maybe could we use this list instead of the one coming from ChatGPT? https://en.wikipedia.org/wiki/List_of_HTTP_header_fields

We may even automate updating the list 😅

dunglas commented:
There is also https://www.iana.org/assignments/http-fields/http-fields.xhtml
It's easy to parse as it's available in XML, but it's only the official headers, and it's not specified if it's a request or a response header.

AlliBalliBaba commented:
I added a script that scrapes all headers from https://github.com/pichillilorenzo/known-http-header-db, but after noticing it hasn't been updated in 3 years, I removed it again. I don't think these headers change very often, maybe PHP will even get a request object before these need to update. I think it's fine to update manually since this is only an optimization.

Instead, I added a test to at least verify that the entered headers are correctly transformed.

AlliBalliBaba commented:
Very weird that there's ASAN issues in go code, could the build cache be corrupted?

AlliBalliBaba commented:
Very strange, cgo apparenty has issues with tests in internal modules.

dunglas commented on cgi.go:
```diff
@@ -158,18 +160,16 @@ func packCgiVariable(key *C.zend_string, value string) C.ht_key_value_pair {
 		}
 
+		// if the header name could not be cached, register it inefficiently
+		k := phpheaders.GetUnCommonHeader(field)
```
Shouldn't we keep otter for this use case? Custom headers are commonly used by web APIs.

AlliBalliBaba replied:
I'll test this again to see how much faster registration with otter is

AlliBalliBaba replied:
I tested this again and otter is indeed a bit faster than concatenating the string as long as the cache doesn't overflow.
I added it back in, but if the need arises it should be fine to get rid of the dependency again since the gain is relatively small.

dunglas commented on frankenphp.go:
```diff
@@ -650,19 +638,16 @@ func go_read_post(threadIndex C.uintptr_t, cBuf *C.char, countBytes C.size_t) (r
-
-	cookies := r.Cookies()
-	if len(cookies) == 0 {
+	cookie := phpThreads[threadIndex].getActiveRequest().Header.Get("Cookie")
```
If two `Cookie` header (or more) are set, only the first one will be passed to PHP.

AlliBalliBaba replied:
Hmm I found this [RFC](https://www.rfc-editor.org/rfc/rfc6265#section-5.4) that states that multiple Cookie Headers are not allowed, but a newer [RFC](https://www.rfc-editor.org/rfc/rfc7540#section-8.1.2.5) states that they should be appended. When testing with Nginx/FPM the Cookie headers will actually also be appended, so I'll adjust the behaviour and add a test 👍

dunglas commented on testdata/server-all-vars-ordered.php:
```diff
@@ -4,17 +4,17 @@
-             'CONTENT_TYPE',
              'DOCUMENT_ROOT',
              'REMOTE_ADDR',
-             'CONTENT_LENGTH',
```
Shouldn't this one be kept?

AlliBalliBaba replied:
I realized it was in there twice, so I removed it once

dunglas commented on cgi.go:
```diff
@@ -194,27 +194,38 @@ func getKnownVariableKeys(thread *phpThread) map[string]*C.zend_string {
+		return h
+	}
+
+	// try to cache cache the header as zend_string
```
```suggestion
	// try to cache the header as zend_string
```

dunglas commented on cgi.go:
```diff
@@ -194,27 +194,38 @@ func getKnownVariableKeys(thread *phpThread) map[string]*C.zend_string {
+	}
+
+	zendHeader := C.frankenphp_init_persistent_string(toUnsafeChar(commonKey), C.size_t(len(commonKey)))
+	thread.knownHeaderKeys[key] = zendHeader
```
Couldn't we populate this cache when booting the thread?

AlliBalliBaba replied:
Hmm I think that might be unnecessary work (more so with dynamic threads). Due to having trusted variable keys, it should already safe time on the first registration.

AlliBalliBaba replied:
Ideally, the cache would directly contain the `zend_strings` once (and not once per thread), but I have not yet found a way to do so.

AlliBalliBaba replied:
I actually found a way to only register them once by explicitly adding the 'interned string' flag on the main thread. I'll test if this is safe, it might also be useful for the known $_SERVER keys.

dunglas replied:
Awesome

dunglas commented:
Thanks!
